### PR TITLE
🐛 fix issues with using koa proxy with other middleware after it

### PR DESCRIPTION
**Pull request by TheMallen**
Fixes an issue where mounting the middleware with another middleware after it would cause the response context to get clobbered and make chrome and curl [return encoding errors. ](https://github.com/nsimmons/koa-better-http-proxy/issues/8)

Also adds some extra checks for correct method, uses the `.json` extension so rails knows what we want from it, and enforces sending the right `Content-Type` header.

### CODE REVIEW

File: packages/koa-shopify-graphql-proxy/src/shopify-graphql-proxy.ts [hunk 1]
```diff
@@ -31,10 +31,17 @@ export default async function shopifyGraphQLProxy(
...
     proxyReqPathResolver() {
       return `https://${shop}${GRAPHQL_PATH}`;
     },
-  })(ctx, next);
+
+    /*
+      We want this middleware to terminate, not fall through to the next in the chain,
+      but sadly it doesn't support not passing a `next` function. To get around this we
+      just pass our own dummy `next` that resolves immediately.
+    */
```
[the remark this answers is not in the document]
Comment (lemonmade) [reply packages/koa-shopify-graphql-proxy/src/shopify-graphql-proxy.ts]: I think the thing that bugs me is that the comment is not actually next to the thing it is commenting. IMO I would put the function call arguments on separate lines and just put this comment overtop of the function returning a promise

File: packages/koa-shopify-graphql-proxy/src/shopify-graphql-proxy.ts [hunk 2]
```diff
@@ -6,7 +6,7 @@ export interface SessionContext extends Context {
 }
 
 export const PROXY_BASE_PATH = '/graphql';
-export const GRAPHQL_PATH = '/admin/api/graphql';
+export const GRAPHQL_PATH = '/admin/api/graphql.json';
```
[the remark this answers is not in the document]
Comment (lemonmade) [reply packages/koa-shopify-graphql-proxy/src/shopify-graphql-proxy.ts]: mentioned IRL but I think the `Accept` header is the more correct way to do this, but I think they work out the same

File: packages/koa-shopify-graphql-proxy/src/shopify-graphql-proxy.ts [hunk 3]
```diff
@@ -31,10 +31,17 @@ export default async function shopifyGraphQLProxy(
...
       return `https://${shop}${GRAPHQL_PATH}`;
     },
-  })(ctx, next);
+
+    /*
+      We want this middleware to terminate, not fall through to the next in the chain,
+      but sadly it doesn't support not passing a `next` function. To get around this we
+      just pass our own dummy `next` that resolves immediately.
+    */
+  })(ctx, () => Promise.resolve());
```
[the remark this answers is not in the document]
Comment (tzvipm) [reply packages/koa-shopify-graphql-proxy/src/shopify-graphql-proxy.ts]: async is just a function that returns a promise, and you can `await` a non-promise, so I think a normal `noop` is fine here.

File: packages/koa-shopify-graphql-proxy/src/test/proxy.test.ts [hunk 4]
```diff
@@ -106,15 +127,31 @@ describe('koa-shopify-graphql-proxy', () => {
...
     await koaShopifyGraphQLProxy(ctx, jest.fn());
 
     const {proxyReqPathResolver} = proxyFactory.mock.calls[0][1];
     expect(proxyReqPathResolver(ctx)).toBe(`https://${shop}${GRAPHQL_PATH}`);
   });
+
+  it('terminates middleware chain when proxying (does not call next)', async () => {
+    const shop = 'some-shop.myshopify.com';
+
+    const ctx = createMockContext({
```
Comment (tzvipm) [inline packages/koa-shopify-graphql-proxy/src/test/proxy.test.ts]: should we define this in the describe as a type of fixture?

File: packages/koa-shopify-graphql-proxy/src/shopify-graphql-proxy.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (TheMallen) [reply packages/koa-shopify-graphql-proxy/src/shopify-graphql-proxy.ts]: Has to be an async noop or TS complains. Found a reasonable way.

File: packages/koa-shopify-graphql-proxy/src/shopify-graphql-proxy.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (TheMallen) [reply packages/koa-shopify-graphql-proxy/src/shopify-graphql-proxy.ts]: image
going with this, I think looking at the way this noop is expressed much further is kinda bike shedding

**Pull request merged**